### DISCUSSION TITLE
unicode chars w pandas , SQL Server and pyodbc

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#10494] [Usage Questions] [answered]
### Describe the bug

When trying to insert a row in a database containing non-latin characters, the results are returned as `???`.

My investigation seems to indicate that this is something that isn't being passed correctly from SQLAlchemy back to the underlying database; specifically its not inserting with `N'加拿大'`.

In debugging it seems to send the characters back to the driver, but the profiled results in SQL Server indicates that its not passing through correctly.

Behaviour happens with ODBC 17 and ODBC 18.

I am also able to insert the characters directly into the database if i execute it in SSMS with the following.

My understanding is it _should_ be sending the unicode down properly to the DB via the driver, but this doesn't seem to be the case.

### SQL to create the table
```sql
CREATE TABLE [oneabc].[stable1](
	[index] [bigint] NULL,
	[sample_char] [nvarchar](max) NULL,
	[c2] [nvarchar](max) NULL
)
```

### Successful insert via SSMS

```sql
INSERT INTO oneabc.stable1 ([index], sample_char, c2)
VALUES (
	20,
	N'加拿大',
	'beta'
)
```

### Python Code to insert row

```python
from sqlalchemy import create_engine

import pandas as pd

ul = 'mssql+pyodbc://sa:yourStrong_Password@127.0.0.1,1433/data?driver=ODBC+Driver+17+for+SQL+Server'
with create_engine(ul).begin() as con:
    (
        pd.DataFrame([['加拿大', 'alpha']], columns=['sample_char', 'c2'])
        .to_sql(
            'stable1', schema='oneabc', con=con, if_exists='append'
        )
    )
```

also tried...

```python
from sqlalchemy import create_engine

import pandas as pd

ul = 'mssql+pyodbc://sa:yourStrong_Password@127.0.0.1,1433/data?driver=ODBC+Driver+17+for+SQL+Server'
with create_engine(ul).begin() as con:
    (
        pd.DataFrame([['加拿大', 'alpha']], columns=['sample_char', 'c2'])
        .assign(
            sample_char = lambda x: x['sample_char'].astype('string')
        )
        .to_sql(
            'stable1', schema='oneabc', con=con, if_exists='append'
        )
    )

```

Same behaviour using `index=False`

### ODBC Logs (Windows)
[sql_trace.zip](https://github.com/sqlalchemy/sqlalchemy/files/13026629/sql_trace.zip)

Two versions; one with the first code, one with the `astype('string')` code. 

### SQL Trace from Sql Server (Linux - 2022 - Docker)

```sql
declare @p1 int
set @p1=NULL
exec sp_prepexec @p1 output,N'@P1 int,@P2 varchar(max),@P3 varchar(max)',N'INSERT INTO oneabc.stable1 ([index], sample_char, c2) VALUES (@P1, @P2, @P3)',0,'???','alpha'
select @p1
```

### Inserted Row Results

| index | sample_char | c2 |
| ---- | --- | --- |
| 0 | ??? | alpha |

### Expected Row Results

| index | sample_char | c2 |
| ---- | --- | --- |
| 0 | 加拿大 | alpha |

### SQLAlchemy Version in Use

2.0.22

### DBAPI (i.e. the database driver)

pyodbc

### Database Vendor and Major Version

SQL Server 2022

### Python Version

3.11

### Operating system

Windows/Linux

### To Reproduce

```python
ul = 'mssql+pyodbc://sa:yourStrong_Password@127.0.0.1,1433/data?driver=ODBC+Driver+17+for+SQL+Server'
with create_engine(ul).begin() as con:
    (
        pd.DataFrame([['加拿大', 'alpha']], columns=['sample_char', 'c2'])
        .to_sql(
            'stable1', schema='oneabc', con=con, if_exists='append'
        )
    )
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
``` 
[sql_trace.zip](https://github.com/sqlalchemy/sqlalchemy/files/13026629/sql_trace.zip)

### Additional context

Appreciate the help!

**Comment by zzzeek** [comment]
hi - 

SQLAlchemy passes through Python unicode strings to pyodbc directly.  It's up to the pyodbc driver to handle these, and we have many many tests to ensure this works.

on your end you would need to ensure the encoding of your database is configured appropriately to accommodate these characters, as well as ensure that pandas is creating the appropriate datatypes which would be NVARCHAR (see https://learn.microsoft.com/en-us/sql/t-sql/data-types/nchar-and-nvarchar-transact-sql?view=sql-server-ver16).  if it's creating VARCHAR then things may get truncated.   from a SQLAlchemy perspective this means they must use the [Unicode](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Unicode) datatype.

run the example below as a demonstration:

```py

from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy import Unicode
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "test_a_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str] = mapped_column(Unicode())


e = create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes", echo='debug')
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add(A(data='加拿大'))
    sess.commit()

with Session(e) as sess:
    value = sess.scalar(select(A.data))

    assert value == '加拿大'


```

**Reply by ghost** [reply]
Perhaps this becomes more of a pandas question.

Seems like everything should be set for unicode, and the characters are correct, but something isn't getting passed down properly to the underlying driver.

Would you suggest opening an issue with the pyodbc or pandas group to move this forward?

Thanks for the reply on this too, appreciate the help.

**Reply by ghost** [reply]
Update as well, i had success specifying `NVARCHAR` on the dtypes for the `.to_sql` function in pandas. Seems like if i'm explicit, it should insert as expected.

Thanks again!

**Reply by ghost** [reply, marked as answer]
```python
from sqlalchemy import create_engine
from sqlalchemy.types import NVARCHAR

import pandas as pd

ul = 'mssql+pyodbc://sa:yourStrong_Password@127.0.0.1,1433/data?driver=ODBC+Driver+17+for+SQL+Server'
with create_engine(ul).begin() as con:
    (
        pd.DataFrame([['加拿大', 'alpha']], columns=['sample_char', 'c2'])
        .assign(
            sample_char = lambda x: x['sample_char'].astype('string')
        )
        .to_sql(
            'stable', schema='oneabc', con=con, if_exists='append', index=False,
            dtype={'sample_char': NVARCHAR, "c2": NVARCHAR}
        )
    )
```

**Reply by gordthompson** [reply]
https://github.com/pandas-dev/pandas/issues/35627
